### PR TITLE
Add ClientFactory::per_host_clients

### PR DESCRIPTION
**Pull request by sfackler**
## Before this PR
Services that wanted to manage node selection themselves had to manually manage client construction and track updates.

## After this PR

==COMMIT_MSG==
Added APIs to create separate clients for each replica of a service.
==COMMIT_MSG==

### CODE REVIEW

File: conjure-runtime/src/per_host_clients.rs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+impl<T> PartialEq for PerHostClients<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.clients.len() == other.clients.len()
+            && self.clients.keys().all(|k| other.clients.contains_key(k))
+    }
+}
+
+impl<T> PerHostClients<T> {
```
Comment (sfackler) [inline conjure-runtime/src/per_host_clients.rs:18]: The API here is very minimal currently - I expect we'll probably expand it in the future but this is enough to support basic workflows.

File: conjure-runtime/src/per_host_clients.rs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+use std::collections::{hash_map, HashMap};
+
+use url::Url;
+
+/// A collection of clients for each host of a replicated service.
+pub struct PerHostClients<T> {
+    pub(crate) clients: HashMap<Host, T>,
+}
+
+impl<T> PartialEq for PerHostClients<T> {
```
Comment (sfackler) [inline conjure-runtime/src/per_host_clients.rs:11]: This impl is a bit jank - we need it for the Refreshable to work, but it's only correct when comparing PerHostClients for a single `per_host_clients` "lineage".

We could add some kind of disambiguator field but I don't think people will end up trying to compare instances in practice.

File: conjure-runtime/src/per_host_clients.rs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    type IntoIter = Iter<'a, T>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter()
+    }
+}
+
+/// An identifier of a single host of a replicated service.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Host {
```
Comment (sfackler) [inline conjure-runtime/src/per_host_clients.rs:39]: I defined this rather than just having a map from Uri to T to support #199.

**Review by TheKeveloper — APPROVED** [review]

File: conjure-runtime/src/client_factory.rs [hunk 4]
```diff
@@ -365,4 +341,129 @@ impl ClientFactory {
...
+
+                    for (i, uri) in c.uris().iter().enumerate() {
+                        let host = Host { uri: uri.clone() };
+                        let host_config = service_config::Builder::from(c.clone())
+                            .uris([uri.clone()])
+                            .build();
+
+                        let state = state_builder.build(&host_config, Some(i))?;
+                        let state = Arc::new(ArcSwap::new(state));
+                        let client = make_client(Client::new(state, None));
```
Comment (sfackler) [inline conjure-runtime/src/client_factory.rs:415]: One implication of this setup that I want to call out is that each individual client is not live reloading - changes to e.g. the read timeout will result in new client objects created but the old ones will not update.

We could adjust this logic to make the clients reload internally but I don't know if it's worth the extra complexity.

**Pull request merged**